[AOTI] Improve the two-pass wrapper codegen

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113762

Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change removes kernel scheduling inconsistency between the two passes and also speeds up the compilation. Another future improvement can be done is to make the second-pass reuse the scheduled nodes and do the wrapper codegen only.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D51405702](https://our.internmc.facebook.com/intern/diff/D51405702)

desertfire committed:
[AOTI] Improve the two-pass wrapper codegen

Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change not only speeds up the compilation, but also removes kernel scheduling inconsistency between the two passes. Another future improvement is to make the second-pass reuse the scheduler and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51405702).

muchulee8 commented on test/inductor/test_aot_inductor.py:
```diff
@@ -1177,6 +1214,7 @@ class AOTInductorTestABICompatibleCpu(TestCase):
         # TODO: test_freezing_abi_compatible_cpu somehow fails on CI but not locally,
         #   NotImplementedError: Cannot access storage of OpaqueTensorImpl
         "test_freezing": TestFailure(("abi_compatible_cpu",), is_skip=True),
+        "test_missing_cubin": TestFailure(("abi_compatible_cpu",)),
```
What will happen if we run cpu tests? It is true that the main purpose is to store/load cubin in different stage, but in theory it should still work fine if we run the exact CPU test?

desertfire replied:
We need to add abi_compatible support for convolution, and the issue is more than adding a missing shim interface. I will add a comment, and come back to fix it later.

muchulee8 commented on test/inductor/test_aot_inductor.py:
```diff
@@ -1202,6 +1240,7 @@ class AOTInductorTestABICompatibleCuda(TestCase):
     # test_failures, xfail by default, set is_skip=True to skip
     {
         "test_dup_unbacked_sym_decl": TestFailure(("abi_compatible_cuda",)),
+        "test_missing_cubin": TestFailure(("abi_compatible_cuda",)),
```
Can we add a comment on what happened? The `# There is a double-free issue which will be fixed in another PR` at the bottom is pretty informative, and people know it will be fixed. It's hard to tell for other tests whether the XFAIL is in purpose or it's need to be fixed?

muchulee8 commented on torch/_inductor/graph.py:
```diff
@@ -974,10 +1019,11 @@ def compile_to_module(self):
             from .codecache import AotCodeCache
 
-            code, linemap = self.codegen()
+            assert self.cpp_wrapper, "AOT mode only supports C++ wrapper"
```
Nice!

chenyang78 commented on torch/_inductor/compile_fx.py:
```diff
@@ -590,8 +516,11 @@ def fx_codegen_and_compile(
+        # and for the forward pass, we have the real inputs to be used as example_inputs.
         graph = GraphLowering(
             gm,
+            example_inputs=V.real_inputs if is_inference else example_inputs,
```
nit - Looks like for the non-forward pass, ```example_inputs``` are faked tensors or sym values. Could you add a comment to make it clear? Sometimes, it's hard to tell if an example_input is of FakedTensor or not by reading the code. Thanks.

chenyang78 reviewed (approved):
Left a minor comment. LGTM, Thanks!

desertfire committed:
Update on "[AOTI] Improve the two-pass wrapper codegen"


Summary: For the second-pass, we don't have to rerun the whole inductor flow again. This PR moves that second-pass to the codegen time. This change removes kernel scheduling inconsistency between the two passes and also speeds up the compilation. Another future improvement can be done is to make the second-pass reuse the scheduled nodes and do the wrapper codegen only.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51405702](https://our.internmc.facebook.com/intern/diff/D51405702)

[ghstack-poisoned]
